temporian/implementation/numpy/operators/set_index.py

```diff
@@ -0,0 +1,186 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import numpy as np
+
+from temporian.core.operators.set_index import SetIndexOperator
+from temporian.implementation.numpy.data.event import NumpyEvent
+from temporian.implementation.numpy.data.event import NumpyFeature
+from temporian.implementation.numpy.data.sampling import NumpySampling
+from temporian.implementation.numpy.operators.utils import _sort_by_timestamp
+
+
+def _append_impl(event: NumpyEvent, append_feat_names: List[str]) -> NumpyEvent:
+    """
+    Append the specified feature names to the index of the given NumpyEvent.
+
+    Args:
+        event:
+            The input NumpyEvent object.
+        append_feat_names:
+            A list of feature names to append to the index.
+
+    Returns:
+        The resulting NumpyEvent object with the updated index.
+    """
+    # destination index names
+    dst_idx_names = event.sampling.index + append_feat_names
+
+    # positions of features that are going to be part of the destination index
+    dst_idx_pos = [
+        event.feature_names.index(append_feat_name)
+        for append_feat_name in append_feat_names
+    ]
+    # positions of features that are being kept
+    dst_feat_pos = {
+        feat_name: pos
+        for pos, feat_name in enumerate(event.feature_names)
+        if feat_name not in append_feat_names
+    }
+    # initialize destination event & sampling data
+    dst_event_data = {}
+    dst_samp_data = {}
+    for src_idx_lvl, src_idx_lvl_feats in event.data.items():
+        dst_idx_suffs = [
+            tuple(x)
+            for x in zip(*[src_idx_lvl_feats[i].data for i in dst_idx_pos])
+        ]
+        for target_dst_idx_suff in set(dst_idx_suffs):
+            # find all occurrences of destination suff in source event
+            dst_idx_suff_pos = [
+                idx
+                for idx, dst_idx_suff in enumerate(dst_idx_suffs)
+                if dst_idx_suff == target_dst_idx_suff
+            ]
```

Comment:
I "feel" creating `dst_idx_suffs` + creating `dst_idx_suff_pos` could be done in the same pass through the `src_idx_lvl_feats[i].data` array - but its hard to say for sure just from looking at the code. Leaving the thought here in case it makes sense and you want to give it a go.

Reply:
it is but you would also need another loop to count the number of elements of each suffix, so that you can later allocate the appropriate memory for each of them